include/pmacc/lockstep/ForEach.hpp

```diff
@@ -120,26 +128,31 @@ namespace pmacc
              */
             template<
                 typename T_Functor,
-                // check if functor is invocable
-                typename = Void<decltype(alpaka::core::declval<T_Functor>()(alpaka::core::declval<Idx const>()))>,
+                // check if functor is invokable
+                typename = IsCallableWithIndex<T_Functor>,
                 std::
                     enable_if_t<std::is_void<InvokeResult_t<T_Functor, Idx const>>::value && domainSize != 1, int> = 0>
             HDINLINE void operator()(T_Functor&& functor) const
             {
-                for(uint32_t i = 0u; i < numCollIter; ++i)
+                static constexpr uint32_t numFullLoops = domainSize / (simdSize * numWorkers);
+                for(uint32_t i = 0u; numFullLoops != 0u && i < numFullLoops; ++i)
                 {
                     uint32_t const beginWorker = i * simdSize;
                     uint32_t const beginIdx = beginWorker * numWorkers + simdSize * this->getWorkerIdx();
-                    if(outerLoopCondition || !innerLoopCondition || beginIdx < domainSize)
+                    for(uint32_t s = 0u; s < simdSize; ++s)
+                        functor(Idx(beginIdx + s, beginWorker + s));
+                }
+
+                // process left over indices if the domainSize is not a multiple of 'simdSie * numWorkers'
+                static constexpr bool hasRemainder = (domainSize % (simdSize * numWorkers)) != 0u;
+                static constexpr uint32_t leftOverIndices = domainSize - (numFullLoops * numWorkers * simdSize);
```

Comment:
Why are these variables static?

Reply:
They should not be static. It is a copy passed bug because I had the definition during my development inside the class.